lib/sync/fullsync.go

```diff
@@ -0,0 +1,547 @@
+	}
+
+	currentTarget := f.peers.getTarget()
+	bestBlockHeader, err := f.blockState.BestBlockHeader()
+	if err != nil {
+		return nil, fmt.Errorf("while getting best block header")
+	}
+
+	// our best block is equal or ahead of current target.
+	// in the nodes pov we are not legging behind so there's nothing to do
+	if uint32(bestBlockHeader.Number) >= currentTarget {
+		return nil, nil
+	}
+
+	startRequestAt := bestBlockHeader.Number + 1
+	targetBlockNumber := startRequestAt + 128
+
+	if targetBlockNumber > uint(currentTarget) {
+		targetBlockNumber = uint(currentTarget)
+	}
+
+	ascendingBlockRequests := network.NewAscendingBlockRequests(startRequestAt, targetBlockNumber,
+		network.BootstrapRequestData)
+
+	return f.createTasks(ascendingBlockRequests), nil
+}
+
+func (f *FullSyncStrategy) createTasks(requests []*network.BlockRequestMessage) []*syncTask {
+	tasks := make([]*syncTask, len(requests))
+	for idx, req := range requests {
+		tasks[idx] = &syncTask{
+			request:      req,
+			response:     &network.BlockResponseMessage{},
+			requestMaker: f.reqMaker,
+		}
+	}
+	return tasks
+}
+
+func (f *FullSyncStrategy) IsFinished(results []*syncTaskResult) (bool, []Change, []peer.ID, error) {
+	repChanges, peersToIgnore, validResp := validateResults(results, f.badBlocks)
+
+	validBlocksUnderFragment := func(highestFinalizedNumber uint, fragmentBlocks []*types.BlockData) []*types.BlockData {
+		startFragmentFrom := -1
+		for idx, block := range fragmentBlocks {
+			if block.Header.Number > highestFinalizedNumber {
+				startFragmentFrom = idx
+				break
+			}
+		}
+
+		if startFragmentFrom < 0 {
+			return nil
+		}
+
+		return fragmentBlocks[startFragmentFrom:]
+	}
+
+	highestFinalized, err := f.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return false, nil, nil, fmt.Errorf("getting highest finalized header")
+	}
+
+	readyBlocks := make([][]*types.BlockData, 0, len(validResp))
+	for _, reqRespData := range validResp {
+		// if Gossamer requested the header, then the response data should
+		// contains the full bocks to be imported
+		// if Gossamer don't requested the header, then the response shoul
+		// only contains the missing parts the will complete the unreadyBlocks
+		// and then with the blocks completed we should be able to import them
+
+		if reqRespData.req.RequestField(network.RequestedDataHeader) {
+			updatedFragment, ok := f.unreadyBlocks.updateDisjointFragments(reqRespData.responseData)
+			if ok {
+				validBlocks := validBlocksUnderFragment(highestFinalized.Number, updatedFragment)
+				if len(validBlocks) > 0 {
+					readyBlocks = append(readyBlocks, validBlocks)
+				}
+			} else {
+				readyBlocks = append(readyBlocks, reqRespData.responseData)
+			}
+			continue
+		}
+
+		completedBlocks := f.unreadyBlocks.updateIncompleteBlocks(reqRespData.responseData)
+		readyBlocks = append(readyBlocks, completedBlocks)
+	}
+
+	// disjoint fragments are pieces of the chain that could not be imported rn
```

Comment:
```suggestion
	// disjoint fragments are pieces of the chain that could not be imported right now
```
I like it but probably should make it clear lol